### PR TITLE
[core][metric] Adding object_store_memory related metrics [1/n] 

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?

This PR adds the metric for:
- object_store_memory by Location:
  - InMemory (shared memory)
  - Spilled
  - ~~Fallback~~ (TODO) 

TODO:
- Tracking the current fallback object size will be in another PR. 
- Some e2e release tests with grafna UI to see how this actually looks like

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_metrics_agent.py [hunk 1]
```diff
@@ -31,9 +31,13 @@
 # NOTE: Commented out metrics are not available in this test.
 # TODO(Clark): Find ways to trigger commented out metrics in cluster setup.
 _METRICS = [
+    # TODO(rickyx): refactoring the below 3 metric seem to be a bit involved
+    # , e.g. need product side changes on the metric subscribers.
     "ray_object_store_available_memory",
```
Comment (rickyyx) [inline python/ray/tests/test_metrics_agent.py:36]: Looks like we used to assume no tags on such metrics. And product has some subscribers here: https://github.com/ray-project/ray/blob/1ce0ab7b7cc6a111140e3f54f02c02725e5cc647/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-agent-config.json#L37-L39

I am not sure how product could support metrics with tags. cc @alanwguo if you know any pointers?

Comment (ericl) [reply python/ray/tests/test_metrics_agent.py:36]: Can you avoid changing old metrics and only add the new resource one? We can deprecate and remove those at a later time.

File: src/ray/stats/metric_defs.h [hunk 2]
```diff
@@ -81,16 +81,18 @@ DECLARE_stats(scheduler_failed_worker_startup_total);
...
 DECLARE_stats(spill_manager_request_total);
 DECLARE_stats(spill_manager_throughput_mb);
 
 /// GCS Storage
 DECLARE_stats(gcs_storage_operation_latency_ms);
 DECLARE_stats(gcs_storage_operation_count);
 
+/// Object Store
+DECLARE_stats(object_store_memory_bytes);
+DECLARE_stats(object_store_memory_count);
```
Comment (ericl) [inline src/ray/stats/metric_defs.h:94]: Let's remove the count for now.

File: src/ray/stats/metric_defs.cc [hunk 3]
```diff
@@ -196,6 +185,43 @@ DEFINE_stats(gcs_storage_operation_count,
...
+             /// Location:
+             ///    - InMemory: currently in memory (ObjectManager)
+             ///    TODO(rickyx): current spilled
+             ///    - Spilled: spilled to disk (LocalObjectManager)
+             ///    - Fallback: fallback allocated (ObjectManager)
+             /// Type:
+             ///     reported from LocalObjectManager
+             ///    - PrimaryCopy: pinned by local raylet and pending spills
+             ///    - PendingSpill: still in memory but to be spilled
+             ///    - PendingRestore: to be restored from spill
```
Comment (ericl) [inline src/ray/stats/metric_defs.cc:204]: PendingRestore shouldn't be reported right? Since that's not actual memory usage.

File: src/ray/stats/metric_defs.cc [hunk 4]
```diff
@@ -196,6 +185,43 @@ DEFINE_stats(gcs_storage_operation_count,
...
+/// also overlap.
+DEFINE_stats(object_store_memory_bytes,
+             "Object store memory by various sub-kinds on this node",
+             /// Location:
+             ///    - InMemory: currently in memory (ObjectManager)
+             ///    TODO(rickyx): current spilled
+             ///    - Spilled: spilled to disk (LocalObjectManager)
+             ///    - Fallback: fallback allocated (ObjectManager)
+             /// Type:
+             ///     reported from LocalObjectManager
```
Comment (ericl) [inline src/ray/stats/metric_defs.cc:201]: I'm a little confused. These states are all for spilled objects only?

File: src/ray/stats/metric_defs.cc [hunk 5]
```diff
@@ -196,6 +185,43 @@ DEFINE_stats(gcs_storage_operation_count,
              (),
              ray::stats::COUNT);
 
+/// Object store
+/// Local object store memory of different types reported from multiple
+/// components. Reported metric from different subtypes, e.g. Location
+/// and Type could overlap. Reported metric within the same type might
```
Comment (ericl) [inline src/ray/stats/metric_defs.cc:191]: What do you mean by overlap? The bytes should always sum up to the total number of actual bytes in memory + disk, right?

**Review by ericl — COMMENTED** [review]
Can you attach some screenshots of how the breakdown looks like for some common workloads? We should also have unit tests that check the breakdowns are correct, similar to test_task_metrics.py

Let's also discuss offline how the breakdowns are created, I have some questions on their meanings.

File: src/ray/stats/metric_defs.cc (hunk 3 above)
Comment (rickyyx) [reply src/ray/stats/metric_defs.cc:204]: Yeah - they do not correspond to a physical chunk of usage, but more of a logical type.

File: src/ray/stats/metric_defs.cc (hunk 4 above)
Comment (rickyyx) [reply src/ray/stats/metric_defs.cc:201]: Oh no, I think the relevant spilled objects are: 
- Spilled (current spilled size on external storage, WIP) 
- PendingRestore (the logical type, no physical memory back up as you mentioned in another comment)
- PendingSpill (still in memory)

The rest are not really about spilled objects.

File: src/ray/stats/metric_defs.cc (hunk 5 above)
Comment (rickyyx) [reply src/ray/stats/metric_defs.cc:191]: Oh I mean overlapping of bytes from different subgroups.  
1. The definition of the metric group: 
For example, the amount of spillable/evictable/inUse bytes overlap with bytes InMemory ; bytes in PrimaryCopy overlap with bytes in "InUse" and "Spillable". 

2. The reporting sources:
So metrics are exported at 3 places now:  the plasma store backend, the LocalObjectManager (spilling mostly) and the ObjectManager. It will probably require some sort of refactoring to centralize it. 


But yeah, I agree some of the groups should add up to 1, for example
- Total usage = `InMemory` + `Fallback` + `Spilled` 
- `InMemory` = `PrimaryCopy` + `PendingSpill` 
- all bytes from different `Creator` (e.g. `Worker`/`RemoteRaylet`) across nodes should sum up to the total size of `InMemory` + `Spilled` ?

I am 100% sure about these though, cc @stephanie-wang ^

**Comment by rkooo567** [comment]
Lmk when this is re-reviewable!

File: python/ray/tests/test_metrics_agent.py [hunk 6]
```diff
@@ -31,9 +31,12 @@
 # NOTE: Commented out metrics are not available in this test.
 # TODO(Clark): Find ways to trigger commented out metrics in cluster setup.
 _METRICS = [
+    # TODO(rickyx): refactoring the below 3 metric seem to be a bit involved
+    # , e.g. need product side changes on the metric subscribers.
```
Comment (ericl) [inline python/ray/tests/test_metrics_agent.py:35]:
```suggestion
```

Comment (ericl) [inline python/ray/tests/test_metrics_agent.py:35]:
```suggestion
    # , e.g. need to see how users currently depend on them.
```

File: python/ray/tests/test_object_store_metrics.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    del objs
+
+    expected = {
+        "InMemory": 0,
+        "Fallback": 0,
+    }
+
+    wait_for_condition(
+        # 1KiB for metadata difference
+        lambda: approx_eq_dict_in(objects_by_loc(info), expected, 1 * KiB),
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/tests/test_object_store_metrics.py:74]: Probably object metadata?

File: python/ray/tests/test_object_store_metrics.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        timeout=20,
+        retry_interval_ms=500,
+    )
+
```
Comment (ericl) [inline python/ray/tests/test_object_store_metrics.py:131]: Add tests for Spilled state too?

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 9]
```diff
@@ -24,6 +24,13 @@ namespace plasma {
 void ObjectStatsCollector::OnObjectCreated(const LocalObject &obj) {
   const auto kObjectSize = obj.GetObjectInfo().GetObjectSize();
   const auto kSource = obj.GetSource();
+  const auto &kAllocation = obj.GetAllocation();
+
+  if (kAllocation.fallback_allocated) {
+    num_bytes_on_fallback_ += kObjectSize;
+  } else {
+    num_bytes_on_shared_memory_ += kObjectSize;
+  }
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:33]: Consider using CounterMap, and registering a callback that records metrics on change.

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 10]
```diff
@@ -169,7 +183,10 @@ void ObjectStatsCollector::OnObjectRefDecreased(const LocalObject &obj) {
 }
 
 void ObjectStatsCollector::RecordMetrics() const {
-  // TODO(sang): Add metrics.
+  ray::stats::STATS_object_store_memory.Record(num_bytes_on_shared_memory_,
+                                               {{"Location", "InMemory"}});
+  ray::stats::STATS_object_store_memory.Record(num_bytes_on_fallback_,
+                                               {{"Location", "Fallback"}});
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.cc:189]: Shall we use capitalized constants? And put these as constants?

File: src/ray/object_manager/plasma/plasma_allocator.cc [hunk 11]
```diff
@@ -89,10 +89,12 @@ absl::optional<Allocation> PlasmaAllocator::Allocate(size_t bytes) {
     return absl::nullopt;
   }
   allocated_ += bytes;
-  return BuildAllocation(mem, bytes);
+  return BuildAllocation(mem, bytes, /* is_fallback_allocated */ false);
 }
```
Comment (ericl) [inline src/ray/object_manager/plasma/plasma_allocator.cc:93]: I thought that previously we decided plasma should only record Unsealed objects, and LocalObjectManager should record locations for the rest (Spilled, InMem)?

File: src/ray/raylet/local_object_manager.h [hunk 12]
```diff
@@ -160,6 +160,10 @@ class LocalObjectManager {
   /// bytes used by objects currently being spilled.
   int64_t GetPrimaryBytes() const;
 
+  /// Get the current number of objects used by primary object copies.
+  /// This number includes  objects currently being spilled.
+  size_t GetPrimaryCount() const;
```
Comment (ericl) [inline src/ray/raylet/local_object_manager.h:165]: Remove Primary / etc tracking code in this PR?

File: src/ray/stats/metric_defs.cc [hunk 13]
```diff
@@ -196,6 +196,18 @@ DEFINE_stats(gcs_storage_operation_count,
              (),
              ray::stats::COUNT);
 
+/// Object store
+// TODO:
+DEFINE_stats(object_store_memory,
+             "Object store memory by various sub-kinds on this node",
+             /// Location:
+             ///    - InMemory: currently in shared memory(e.g. /dev/shm)
+             ///    - Spilled: spilled to external storage
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/stats/metric_defs.cc:205]: I think it's important we include spilling this in this PR, since it impacts the design of the other recording points. Can we actually punt fallback allocated to another PR (just treat it as InMemory?)

File: src/ray/object_manager/plasma/plasma_allocator.cc (hunk 11 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/plasma_allocator.cc:93]: Yeah - after chatting with Stephanie, the InMem info on LocalObjectManager is incomplete: LocalObjectManager only sees the Primary Copy related usage (those pinned). The total InMem needs to be reported from the plasma. 

So the current proposed split will be: 
- LocalObjectManager: Spill bytes
- Plasma store: Fallback + InMemory

Comment (ericl) [reply src/ray/object_manager/plasma/plasma_allocator.cc:93]: Ok, that makes sense.

File: src/ray/raylet/local_object_manager.cc [hunk 14]
```diff
@@ -217,8 +218,6 @@ bool LocalObjectManager::SpillObjectsOfSize(int64_t num_bytes_to_spill) {
           auto now = absl::GetCurrentTimeNanos();
           RAY_LOG(DEBUG) << "Spilled " << bytes_to_spill << " bytes in "
                          << (now - start_time) / 1e6 << "ms";
-          spilled_bytes_total_ += bytes_to_spill;
```
Comment (rickyyx) [inline src/ray/raylet/local_object_manager.cc:220]: Moved this from the callback to `OnObjectsSpilled` - I don't think we were tracking spilled size updates from functions like `SpillObjects`.

File: src/ray/raylet/local_object_manager.cc [hunk 15]
```diff
@@ -539,6 +544,11 @@ void LocalObjectManager::ProcessSpilledObjectsDeleteQueue(uint32_t max_batch_siz
         object_urls_to_delete.emplace_back(object_url);
       }
       spilled_objects_url_.erase(spilled_objects_url_it);
+
+      // Update current spilled objects metrics
+      RAY_CHECK(local_objects_.contains(object_id))
+          << "local objects might should contain the spilled object: " << object_id;
```
Comment (scv119) [inline src/ray/raylet/local_object_manager.cc:550]: nit: local objects should contain the spilled object

File: src/ray/raylet/main.cc [hunk 16]
```diff
@@ -213,7 +213,7 @@ int main(int argc, char *argv[]) {
         node_manager_config.report_resources_period_ms =
             RayConfig::instance().raylet_report_resources_period_milliseconds();
         node_manager_config.record_metrics_period_ms =
-            RayConfig::instance().metrics_report_interval_ms() / 2;
+            RayConfig::instance().metrics_report_interval_ms();
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/raylet/main.cc:216]: You should divide it by 2 to avoid any weird sampling artifacts. Can we avoid changing these quantities if possible in larger PRs?

File: src/ray/object_manager/plasma/stats_collector.h [hunk 17]
```diff
@@ -60,6 +60,8 @@ class ObjectStatsCollector {
  private:
   friend struct ObjectStatsCollectorTest;
 
+  int64_t num_bytes_in_memory_ = 0;
```
Comment (ericl) [inline src/ray/object_manager/plasma/stats_collector.h:63]: Do we need a new counter? It seems you can derive this from some of the counters below.

File: src/ray/object_manager/plasma/store.cc [hunk 18]
```diff
@@ -554,8 +556,15 @@ void PlasmaStore::PrintAndRecordDebugDump() const {
...
-  // TODO(sang): Add metrics.
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Exporting plasma store metrics.";
   object_lifecycle_mgr_.RecordMetrics();
+
+  // Schedule the next metric export callback
+  metrics_timer_ = execute_after(
+      io_context_,
+      [this]() { RecordMetrics(); },
+      RayConfig::instance().metrics_report_interval_ms());
```
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:567]: Divide by 2.

File: src/ray/object_manager/plasma/store.cc [hunk 19]
```diff
@@ -554,8 +556,15 @@ void PlasmaStore::PrintAndRecordDebugDump() const {
 }
 
 void PlasmaStore::RecordMetrics() const {
-  // TODO(sang): Add metrics.
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Exporting plasma store metrics.";
   object_lifecycle_mgr_.RecordMetrics();
+
+  // Schedule the next metric export callback
+  metrics_timer_ = execute_after(
```
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:564]: We should setup this once, not each iteration of RecordMetrics, right?

File: src/ray/object_manager/plasma/store.cc [hunk 20]
```diff
@@ -118,6 +118,9 @@ PlasmaStore::PlasmaStore(instrumented_io_context &main_service,
   if (event_stats_print_interval_ms > 0 && RayConfig::instance().event_stats()) {
     PrintAndRecordDebugDump();
   }
+  if (RayConfig::instance().metrics_report_interval_ms() > 0) {
+    RecordMetrics();
+  }
```
Comment (ericl) [inline src/ray/object_manager/plasma/store.cc:123]: Can we revert this change?

File: src/ray/object_manager/plasma/store.cc [hunk 21]
```diff
@@ -545,7 +548,6 @@ bool PlasmaStore::IsObjectSpillable(const ObjectID &object_id) {
 
 void PlasmaStore::PrintAndRecordDebugDump() const {
   absl::MutexLock lock(&mutex_);
```
[the remark this answers is not in the document]
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.cc:550]: This will associate the metric export with `event_stats_print_interval_ms` rather than the `metrics_report_interval_ms` ?

File: src/ray/object_manager/plasma/store.cc (hunk 19 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.cc:564]: I believe this one works kind of recursively - I followed the pattern of the `stats_timer_` above mainly for consistency. We could also set up a periodic runner like you mention.

File: python/ray/tests/test_object_store_metrics.py [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+
+def test_spilling(object_spilling_config, shutdown_only):
```
Comment (rickyyx) [inline python/ray/tests/test_object_store_metrics.py:95]: Any idea on how to test the unsealed memory tracking specifically? 
It seems to me the create then seal path only happens when we restore spilled objects from storage directly. So one potential way I could test it is by mocking an external system with slow data copy. 

But I also feel that if the in memory tracking works properly, the unseal tracking should be the same since it's really similar code paths.

Comment (rkooo567) [reply python/ray/tests/test_object_store_metrics.py:95]: it's probably tricky to do e2e. Maybe we can do at unit test level in cpp?

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 23]
```diff
@@ -26,6 +27,8 @@ void ObjectStatsCollector::OnObjectCreated(const LocalObject &obj) {
   const auto kSource = obj.GetSource();
 
   num_bytes_created_total_ += kObjectSize;
+  // TODO(rickyx):
+  // Add fallback memory accounting here.
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/stats_collector.cc:31]: is it going to be in a follow up PR?

File: src/ray/object_manager/plasma/stats_collector.cc [hunk 24]
```diff
@@ -168,8 +171,22 @@ void ObjectStatsCollector::OnObjectRefDecreased(const LocalObject &obj) {
...
+int64_t ObjectStatsCollector::GetNumBytesCreatedCurrent() const {
+  return num_bytes_created_by_worker_ + num_bytes_restored_ + num_bytes_received_ +
+         num_bytes_errored_;
+}
+
 void ObjectStatsCollector::RecordMetrics() const {
-  // TODO(sang): Add metrics.
+  ray::stats::STATS_object_store_memory.Record(
+      GetNumBytesCreatedCurrent() - num_bytes_unsealed_,
+      {{"Location", ray::rpc::ObjectLocation_Name(ray::rpc::ObjectLocation::IN_MEMORY)}});
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/stats_collector.cc:182]: Are we planning to track "Used"/"Available"? Isn't the sum of these metrics supposed to be the total memory?

File: src/ray/object_manager/plasma/store.cc [hunk 25]
```diff
@@ -554,8 +556,15 @@ void PlasmaStore::PrintAndRecordDebugDump() const {
...
-  // TODO(sang): Add metrics.
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Exporting plasma store metrics.";
   object_lifecycle_mgr_.RecordMetrics();
+
+  // Schedule the next metric export callback
+  metrics_timer_ = execute_after(
+      io_context_,
+      [this]() { RecordMetrics(); },
+      RayConfig::instance().metrics_report_interval_ms() / 2);
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.cc:567]: Can you comment why?

File: src/ray/object_manager/plasma/store.cc [hunk 26]
```diff
@@ -554,8 +556,15 @@ void PlasmaStore::PrintAndRecordDebugDump() const {
...
 void PlasmaStore::RecordMetrics() const {
-  // TODO(sang): Add metrics.
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Exporting plasma store metrics.";
   object_lifecycle_mgr_.RecordMetrics();
+
+  // Schedule the next metric export callback
+  metrics_timer_ = execute_after(
+      io_context_,
+      [this]() { RecordMetrics(); },
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.cc:566]: Why don't we just call it here? I think `PlasmaStore::RecordMetrics()` must be already periodically called?

File: src/ray/object_manager/plasma/store.h [hunk 27]
```diff
@@ -281,6 +281,9 @@ class PlasmaStore {
   /// Timer for printing debug information.
   mutable std::shared_ptr<boost::asio::deadline_timer> stats_timer_ GUARDED_BY(mutex_);
 
+  /// Timer for metric exporting.
+  mutable std::shared_ptr<boost::asio::deadline_timer> metrics_timer_ GUARDED_BY(mutex_);
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/store.h:285]: Does it have to be shared ptr? maybe it can just be `boost::asio::deadline_timer metrics_timer_`?

File: src/ray/protobuf/common.proto [hunk 28]
```diff
@@ -760,3 +760,15 @@ message NamedActorInfo {
...
+// Currently used for metric exporting only.
+enum ObjectLocation {
+  // TODO(rickyx): seperate fallback memory from this enum
+  // Shared memory sealed + fallback memory sealed
+  IN_MEMORY = 0;
+  // Spilled to external storage
+  SPILLED = 1;
+  // Unsealed memory
+  UNSEALED = 2;
+}
```
Comment (rkooo567) [inline src/ray/protobuf/common.proto:774]: new line!

File: src/ray/raylet/local_object_manager.cc [hunk 29]
```diff
@@ -613,6 +622,18 @@ void LocalObjectManager::RecordMetrics() const {
   ray::stats::STATS_spill_manager_request_total.Record(spilled_objects_total_, "Spilled");
   ray::stats::STATS_spill_manager_request_total.Record(restored_objects_total_,
                                                        "Restored");
+
+  ray::stats::STATS_object_store_memory.Record(
+      spilled_bytes_current_,
+      {{"Location", ray::rpc::ObjectLocation_Name(ray::rpc::ObjectLocation::SPILLED)}});
+}
+
+size_t LocalObjectManager::GetCurrentSpilledCount() const {
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:631]: Is it test only? Can you comment it? (if it is test only, we can also just use friend class)

File: python/ray/tests/test_object_store_metrics.py [hunk 30]
```diff
@@ -0,0 +1,176 @@
...
+        "UNSEALED": 0,
+    }
+    wait_for_condition(
+        # 1KiB for metadata difference
+        lambda: approx_eq_dict_in(objects_by_loc(info), expected, 1 * KiB),
+        timeout=20,
+        retry_interval_ms=500,
+    )
+
+    # Delete all
```
Comment (rkooo567) [inline python/ray/tests/test_object_store_metrics.py:154]: nice test!

**Review by rkooo567 — COMMENTED** [review]
Mostly LGTM

File: src/ray/object_manager/plasma/stats_collector.cc (hunk 23 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/stats_collector.cc:31]: yep

File: src/ray/object_manager/plasma/stats_collector.cc (hunk 24 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/stats_collector.cc:182]: Yes, I guess this will be in a follow-up for 2.2 maybe. More breakdowns of the usage.

File: src/ray/object_manager/plasma/store.cc (hunk 26 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.cc:566]: It was priodically called in `PrintAndRecordDebugDump` which has a similiar timer setup.

File: src/ray/object_manager/plasma/store.cc (hunk 25 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.cc:567]: I guess the metric sampling interval should at least be smaller/equal to the metric reporting interval?

File: src/ray/object_manager/plasma/store.h (hunk 27 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.h:285]: maybe not - i didn't look in great details of why this was this actually.

File: src/ray/raylet/local_object_manager.cc (hunk 29 above)
Comment (rickyyx) [reply src/ray/raylet/local_object_manager.cc:631]: oh yeah - test only for now.

File: python/ray/tests/test_object_store_metrics.py (hunk 22 above)
Comment (rickyyx) [reply python/ray/tests/test_object_store_metrics.py:95]: So for cpp we do have unit test covering the unsealed stats correctness: https://github.com/ray-project/ray/blob/e9755d87a641b02daf82604ef23ce09647158cd4/src/ray/object_manager/plasma/test/stats_collector_test.cc#L75


Which is why I feel specifically testing the unsealed metrics is not super necessary: 
-  the tracking is tested correctly in c++
- the reporting is tested with the in memory memory testing e2e.

File: src/ray/protobuf/common.proto (hunk 28 above)
Comment (rickyyx) [reply src/ray/protobuf/common.proto:774]: there is actually no new line if you hover on the icon.

File: src/ray/object_manager/plasma/store.h (hunk 27 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.h:285]: I agree with you that this is probably not needed - i think Chen had a previous PR thatt touches on some of the data race and the locking here. https://github.com/ray-project/ray/pull/18312

Maybe this could be something to be refactored in a seperate PR - I am leaning towards sticking with the existing code here.

File: src/ray/object_manager/plasma/store.cc (hunk 19 above)
Comment (rickyyx) [reply src/ray/object_manager/plasma/store.cc:564]: Reverting.

File: src/ray/object_manager/plasma/store.cc [hunk 31]
```diff
@@ -113,9 +113,8 @@ PlasmaStore::PlasmaStore(instrumented_io_context &main_service,
                 this->AddToClientObjectIds(object_id, request->client);
               },
           [this](const auto &request) { this->ReturnFromGet(request); }) {
-  const auto event_stats_print_interval_ms =
-      RayConfig::instance().event_stats_print_interval_ms();
-  if (event_stats_print_interval_ms > 0 && RayConfig::instance().event_stats()) {
+  if (RayConfig::instance().event_stats_print_interval_ms() > 0 &&
```
Comment (rickyyx) [inline src/ray/object_manager/plasma/store.cc:116]: Reverted the change to have metrics recording still part of the PrintStats loop. 
So the default `event_stats_print_interval_ms` is 1min, I wonder if we should change the default? In the events `event_stats_print_interval_ms` greater than the metric_reports interval, no metrics would be there. 

If not - maybe we could have another interval?

Comment (rickyyx) [reply src/ray/object_manager/plasma/store.cc:116]: I actually feel we will eventually make a larger change to follow how metrics are exported just like the task metrics, i.e. use something like a counter map and register callback on the change, rather than periodically recording them from member fields. 

This is probably the better way when we add another label, e.g. job id. (Stephanie has requested such a breakdown as multiple jobs might be run at the same time).

File: src/ray/protobuf/common.proto [hunk 32]
```diff
@@ -760,3 +760,15 @@ message NamedActorInfo {
   string ray_namespace = 1;
   string name = 2;
 }
+
+// Info about an object's location.
+// Currently used for metric exporting only.
+enum ObjectLocation {
```
Comment (ericl) [inline src/ray/protobuf/common.proto:766]: Is this actually used as a proto message? Prefer a C++ enum if not.

**Review by ericl — APPROVED** [review]

File: src/ray/protobuf/common.proto (hunk 28 above)
Comment (rickyyx) [reply src/ray/protobuf/common.proto:762]: oh lol there needs to be a new line...

File: src/ray/object_manager/plasma/store.cc (hunk 25 above)
Comment (rkooo567) [reply src/ray/object_manager/plasma/store.cc]: oh I just meant to comment why (I know why lol) haha

File: src/ray/stats/tag_defs.h [hunk 33]
```diff
@@ -45,3 +45,9 @@ extern const TagKeyType WorkerIdKey;
 extern const TagKeyType JobIdKey;
 
 extern const TagKeyType NameKey;
+
+// Object store memory location tag constants
+extern const TagKeyType LocationKey;
+constexpr char kObjectLocInMemory[] = "IN_MEMORY";
+constexpr char kObjectLocSpilled[] = "SPILLED";
+constexpr char kObjectLocUnsealed[] = "UNSEALED";
```
Comment (rkooo567) [inline src/ray/stats/tag_defs.h:53]: newline?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
@rickyyx ping me when you are ready to merge it

**Comment by rickyyx** [comment]
> @rickyyx ping me when you are ready to merge it

Was waiting for some more signals on the build: https://buildkite.com/ray-project/oss-ci-build-pr/builds/1606#_

Ready to merge now!
